Some update to the speed utility

juergenchrist opened the pull request:
The speed utility has some minor knits that cause it to produce varying output depending on the memory subsystem.  This can easily be alleviated by locking the used pages into memory.

t8m commented:
IMO the mlock is rather a feature. The wait-for-children is a bug fix that should be backported IMO. So this should be split into two PRs.

juergenchrist commented:
> IMO the mlock is rather a feature. The wait-for-children is a bug fix that should be backported IMO. So this should be split into two PRs.

I can do that.  I was hoping making one PR with two commits would be sufficient.

juergenchrist commented:
Waiting for children moved to #19093.

paulidale commented:
The code in mem_sec.c does a bit more: https://github.com/openssl/openssl/blob/723844d3762c05727e8f6f21d0a1098e23302ebd/crypto/mem_sec.c#L543-L558

Moreover there is a maze of preprocessor conditionals preventing it's use: more than a few of our target platforms do not support page locking.

juergenchrist commented:
> The code in mem_sec.c does a bit more

I have seen this code.  But the goal here is to have the whole memory in RAM before the throughput measurements start.  So I modelled that code after the mem_sec.c code without the mlock2 case which does not give any benefits for this use case.

juergenchrist commented:
What is missing here?  Should I use the macro `OPENSSL_NO_SECURE_MEMORY`?  I don't think the distinction between mlock and mlock2 is needed in this case.  We anyway do a pre-faulting after locking the memory (via `memset`).  Since all CI-checks have passed it seems to me that all considered architectures are covered.

juergenchrist commented:
Updated to code to only use memory locking on Linux or Windows.  Added an option to allow users of the speed utility to decide whether or not mlock should be used.

juergenchrist commented:
@t8m @paulidale polite ping.  Without this patch, our performance team reports up to 20% deviation in results depending on the size of the buffer.  This patch fixes the situation for them.

openssl-machine commented:
This pull request is ready to merge

paulidale commented:
Merged, thanks for the contribution.
